### DISCUSSION TITLE
Wireguard not available as an option when launching the installer script

### DISCUSSION THREAD

**Discussion by simlocker** [pivpn/pivpn#1548] [Support]
Hi everyone,

Recently I was trying to reinstall Wireguard on my Linux device running Dietpi OS (HW is Orange Pi Zero) but on the installation screen it goes straight to OpenVPN, and I'm not given the option to choose between Wireguard and OpenVPN.

This wasn't something that happened before. I've always been able to choose between Wireguard or OpenVPN when installing PiVPN.

Any clues on why might this be happening?
Thanks

**Comment by coolapso** [comment]
have you actually even read what i have said in my comment and the links that i have pointed you towards?

 https://github.com/pivpn/pivpn/issues/1547#issuecomment-1152175529

**Comment by simlocker** [comment]
Yes.

You provided with this link: https://docs.pivpn.io/faq/#what-boardsoses-does-pivpn-support

It doesn't apply to this issue. I haven't changed my SBC hardware.

Then this: https://github.com/pivpn/pivpn#feedback--support

I'm now using the Discussion board instead of the Issues section.

**Reply by coolapso** [reply]
I think i was very clear. but in case i was not: 

If the answer is not there, provide the data we ask for in the issue template. The data you cannot provide you don't but there is info there that you can actually provide is very important to either me or anyone else that might wanna help you.

**Comment by simlocker** [comment]
Okay, here's the template, try to think for a minute not everyone is a genius at IT, and it's not that easy for everyone to even complete it and follow 100% step by step:


## In raising this issue, I confirm the following:

`{please fill the checkboxes, e.g: [X]}`

- [X] I have read the [documentation](https://docs.pivpn.io)
- [] Is it a feature request? please consider opening a [Discussion] (https://github.com/pivpn/pivpn/discussions/new)
- [x] I have read and understood the [contributors guide](https://github.com/pivpn/pivpn/blob/master/CONTRIBUTING.md).
- [] The issue I am reporting can be *replicated*.  
- [x] The issue I am reporting *is* directly related to the pivpn installer script.
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pivpn/pivpn/wiki/FAQ), [closed issues](https://github.com/pivpn/pivpn/issues?q=is%3Aissue+sort%3Aupdated-desc+is%3Aclosed), and [open issues](https://github.com/pivpn/pivpn/issues?q=is%3Aissue+sort%3Aupdated-desc+is%3Aopen)).



### Has your install failed? 
```
yes
```

### Describe the issue


No Wireguard option available when installing.

**Expected behavior**

Wireguard/OpenVPN should be available to select.



### Can you replicate the issue? Describe the steps below

Steps to reproduce the behavior:

I execute the installer script, and I don't get to the point where Wireguard/OpenVPN are presented. I get straight to OpenVPN configuration.

### Have you searched for similar issues and solutions?

Yes, no solution found.


**Additional context**


### Have you taken any steps towards solving your issue?
```
Uninstall, Reinstall.
```

### Please provide your system information
#### What type of hardware are you running PiVPN at?

```
Orange Pi Zero
```
#### Output of `uname -a`
```
Linux orangedietpi 5.15.43-sunxi #22.05.1 SMP Sat May 28 08:17:47 UTC 2022 armv7l GNU/Linux
```

#### Output of `cat /etc/os-release`
```
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

### If install failed Please provide the console output of  `curl -L https://install.pivpn.io | bash`
```
Install doesn't fail, I don't get Wireguard as an option when trying to install. Therefore, I cannot continue with the installation.
```

**Reply by coolapso** [reply]
> Okay, here's the template, try to think for a minute not everyone is a genius at IT, and it's not that easy for everyone to even complete it and follow 100% step by step:

I could ask you the same ... put yourself in our shoes, we are not supposed to guess what its wrong ... to find where the issue is .. we need data, that's why the template exists. filling it to the best of ones ability is better than not filling it at all. 

This being said the issue is related with the architecture and the OS itself. 

Seems that "orangedietpi" s not providing either a wireguard package or a kernel with builtin wireguard, therefore pivpn installer does not detect it and only offers you openvpn.

Even tho this might have worked in the past, something might have changed in the OS you are using which caused it to not work anymore, and its failing to pass the evaluations here https://github.com/pivpn/pivpn/blob/master/auto_install/install.sh#L486-L517

however would need to look with a bit more attention to the code, or maybe @orazioedoardo or @MichaIng  have a better answer than i do regarding this.

**Reply by coolapso** [reply]
to be honest it really seems that in the end it all boils down to this: https://docs.pivpn.io/faq/#what-boardsoses-does-pivpn-support 

Orange Pi is not on the diet pi list of supported boards, 
Seems that "orangedietpi" is a modified version of dietpi with linux sunxi kernel which probably does not have builtin wireguard.

**Comment by simlocker** [comment]
Okay, thanks for the answer. It's still quite odd that it used to work in the past but after trying to reinstall it's going straight to OpenVPN now.

Orange Pi Zero is indeed not in the official list, but I managed to get it working without issues by following this guide: https://github.com/MichaIng/DietPi/issues/1285#issue-280771944

**Reply by MichaIng** [reply]
I bet it's the legacy original kernel provided by Orange Pi. What do `uname -a` and `dpkg -l | grep linux` say?

If it's the legacy kernel, like Linux 3.x or 4.x and it's installed as dpkg package, you might be able to derive the related kernel headers package by replacing "image" with "headers" in the name of the kernel package. But PiVPN will still not support it that way, so WireGuard would then need to be compiled manually.

We generally recommend in such case to use Armbian images, which ship with recent Linux 5.x with builtin WireGuard module. You can then covert it to DietPi via: https://dietpi.com/docs/hardware/#make-your-own-distribution

The note/warning boxes can be ignored in this case. Armbian mainline kernel images are fully supported. I'll update these.

**Reply by coolapso** [reply]
its in the comment above @MichaIng on the filled template: 

`Linux orangedietpi 5.15.43-sunxi #22.05.1 SMP Sat May 28 08:17:47 UTC 2022 armv7l GNU/Linux`

My guess is that the sunxi kernel was not compiled with the required modules?

**Reply by MichaIng** [reply]
That's the Armbian kernel 🤔. Can you show the output of:
```sh
modinfo wireguard
dpkg -L linux-image-sunxi-current | grep wireguard
```

To be precise, the module is not builtin but shipped with the package as dedicated module. At least it should be.

**Comment by simlocker** [comment]
dpkg -l | grep linux: 
```
ii  console-setup-linux               1.205                          all          Linux specific part of console-setup
ii  libselinux1:armhf                 3.1-3                          armhf        SELinux runtime shared libraries
ii  linux-base                        4.6                            all          Linux image base package
ii  linux-dtb-current-sunxi           22.05.1                        armhf        Armbian Linux DTB, version 5.15.43-sunxi current
ii  linux-image-current-sunxi         22.05.1                        armhf        Linux kernel, armbian version 5.15.43-sunxi current
ii  linux-u-boot-orangepizero-current 22.05.1                        armhf        Uboot loader 2021.10
ii  util-linux                        2.36.1-8+deb11u1               armhf        miscellaneous system utilities
```

modinfo wireguard:
```
filename:       /lib/modules/5.15.43-sunxi/kernel/drivers/net/wireguard/wireguard.ko.xz
alias:          net-pf-16-proto-16-family-wireguard
alias:          rtnl-link-wireguard
version:        1.0.0
author:         Jason A. Donenfeld <Jason@zx2c4.com>
description:    WireGuard secure network tunnel
license:        GPL v2
srcversion:     DCB386D1C4DBE397FBCACDA
depends:        libblake2s,udp_tunnel,curve25519-neon,libchacha20poly1305,ip6_udp_tunnel,libcurve25519-generic
intree:         Y
name:           wireguard
vermagic:       5.15.43-sunxi SMP mod_unload ARMv7 thumb2 p2v8
sig_id:         PKCS#7
signer:         Build time autogenerated kernel key
sig_key:        73:F9:8C:1A:0B:D9:BE:11:96:8D:73:60:AB:B0:C9:D5:49:50:28:16
sig_hashalgo:   sha1
signature:      56:D2:B8:EB:EF:77:10:26:42:1F:EC:5E:3D:69:F8:30:C6:FE:89:A6:
                7C:E4:E4:D6:3D:BA:AC:FB:51:B1:FB:9E:80:74:E7:60:5F:25:26:02:
                38:CA:12:4F:1E:52:20:A7:41:A6:3F:66:A6:C2:AD:0B:21:6E:F0:F0:
                B8:02:3B:F4:FF:98:CD:8E:38:F0:FF:4D:CB:77:0E:B8:46:A2:EB:BD:
                DD:6E:36:8E:03:C9:75:2E:82:50:15:66:CF:E1:08:87:19:15:9E:BA:
                A5:D6:C8:0E:25:27:3F:E3:74:1D:99:DD:73:9E:39:6F:80:45:D6:B7:
                7A:03:9E:4B:4B:2B:1C:63:C6:69:1A:6D:56:E8:2A:CE:7F:EF:05:C5:
                86:6B:60:26:23:4F:BC:CD:BC:63:14:BD:F3:33:05:8D:41:A6:5C:DC:
                07:BB:7E:6A:24:5A:51:B7:85:FF:DE:C2:55:0D:D0:83:51:0F:94:A2:
                13:40:72:F7:29:F0:7D:3D:DC:85:EC:BF:5F:28:4B:AC:B7:07:8A:80:
                C8:94:77:89:A9:CD:BE:CF:35:B2:4F:0C:37:7F:8E:CE:DA:A3:58:47:
                9C:7F:79:CE:25:C5:B8:66:8E:E6:87:37:17:2D:A3:C2:CB:F7:E0:ED:
                89:22:76:66:00:53:44:55:51:DE:41:76:5C:D7:95:29:B2:CE:55:0F:
                EB:8F:F6:54:05:5C:FC:28:C9:71:5E:CE:31:09:A0:8A:A1:5E:AA:09:
                67:F5:35:A4:09:88:27:4A:8C:C5:F3:DA:C3:14:10:E5:D0:09:28:E8:
                94:87:56:65:F4:8B:D3:7F:39:82:79:F9:39:FC:42:9F:2C:4D:A1:A8:
                15:66:9A:BA:62:D8:5F:BF:44:A6:DC:F2:16:E2:62:72:8E:F4:30:D4:
                DB:95:16:81:30:B9:68:5C:52:46:E2:5E:3A:A6:8E:1A:41:08:67:D4:
                8E:34:F1:0E:37:74:0A:FC:AE:C8:F9:08:F7:20:70:13:3C:F1:56:CA:
                1D:DB:A3:C5:83:55:5E:D4:8A:74:51:B5:3B:D1:B2:F6:DE:BF:42:8C:
                72:0B:01:2B:F4:88:BD:62:5A:9F:E8:50:92:96:F3:A5:3E:1C:33:78:
                65:EB:38:7E:71:57:16:4C:CF:E9:02:F9:77:C2:3A:49:F3:BA:A7:F7:
                F7:0F:3F:AB:8D:55:D9:71:3F:C7:9F:81:A0:8F:9E:C3:02:1C:1B:74:
                65:AB:1F:21:8C:A6:B3:F9:B3:6B:9C:44:D4:30:49:26:EC:67:5A:3F:
                21:54:EB:9E:60:2E:9D:61:35:CB:3D:0B:62:29:E1:3C:D3:A0:1C:55:
                A7:87:DD:2B:E6:FC:D4:EF:FE:FE:67:F5

dpkg-query: package 'linux-image-sunxi-current' is not installed
Use dpkg --contents (= dpkg-deb --contents) to list archive files contents.
```


But it still wouldn't explain why was it working before. All I did was reinstall PiVPN, and Wireguard was never offered again.
I'm now using a DietPi backup by the way, and PiVPN Wireguard is running without any issues. But in case I need to reinstall to it, I'm afraid it'd produce the same issue.

**Reply by coolapso** [reply]
Edited your comment so it can be more readable. (wraped the cosole in between ``` ). 

It was probably working before, because if i remember correctly early stage we were actually installing the kernel headers the dkms modules or even compiling them to get them installed ... once it started to be shiped by default with linux kernel we stopped doing so. that is one possibility. 

Also you have to consider that things changed, both the Operative systems and PiVPN get's updated. and when you are not using officially supported hardware/Distros its perfectly normal for things to break.

**Reply by MichaIng** [reply]
Ah, since v5.15 Linux by default ships kernel modules in xz-compressed from. This is not yet detected/supported by PiVPN. Let's push a PR to add support for that.

**Reply by coolapso** [reply]
Aight .. moving this to issues as a bug then

**Reply by coolapso** [reply]
https://github.com/pivpn/pivpn/issues/1549

**Comment by MichaIng** [comment]
PR up to fix/implement this: #1550

**Reply by coolapso** [reply]
going to merge as soon as the tests finish. Might merge it to master right afterwards as well.

**Comment by simlocker** [comment]
Thank you both.

**Reply by coolapso** [reply]
@gaston-arg Try now installing form test branch, instructions here: https://docs.pivpn.io/install/#to-install-from-testdevelopment-branch and let us know if it worked.

**Reply by simlocker** [reply]
No, I'm afraid it's not working. Here are the uninstall and install logs:

```
Step: Uninstalling PiVPN: openvpn/wireguard server install & management tool

[ INFO ] DietPi-Software | pivpn -u, please wait...
::: Preparing to remove packages, be sure that each may be safely removed depending on your operating system.
::: (SAFE TO REMOVE ALL ON RASPBIAN)
::: Do you wish to completely remove PiVPN configuration and installed packages from your system? (You will be prompted for each package) [y/n]: y
::: Stopping and disabling services...
::: Removing firewall rules...
vm.swappiness = 100
::: Purge dependencies...
::: Do you wish to remove wireguard-tools from your system? [Y/n]: y
:::     Removing wireguard-tools...done!
::: Do you wish to remove qrencode from your system? [Y/n]: y
:::     Removing qrencode...done!
::: Auto removing remaining dependencies...done!
::: Auto cleaning remaining dependencies...done!
:::
::: Removing VPN configuration files...
:::
::: Removing pivpn system files...
:::
::: Finished removing PiVPN from your system.
::: Reinstall by simply running
:::
:::     curl -L https://install.pivpn.io | bash
:::
::: at any time!
:::
It is strongly recommended to reboot after un-installation.
Would you like to reboot now? [y/n]: y

 Step: Finalising uninstall

[ INFO ] DietPi-Software | APT autopurge, please wait...
Reading package lists...
Building dependency tree...
Reading state information...
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
[  OK  ] DietPi-Software | APT autopurge
[  OK  ] DietPi-Software | systemctl daemon-reload

 DietPi-Software
─────────────────────────────────────────────────────
 Step: Uninstall completed


root@orangedietpi:~# curl -L https://test.pivpn.io | TESTING= bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   162  100   162    0     0    274      0 --:--:-- --:--:-- --:--:--   275
100 93674  100 93674    0     0  81811      0  0:00:01  0:00:01 --:--:--  219k
:::
::: You are root.
::: Hostname length OK
::: Verifying free disk space...
:::
::: Package Cache update is needed, running apt-get update -y ...
 done!
:::
::: Checking apt-get for upgraded packages.... done!
:::
::: Your system is up to date! Continuing with PiVPN installation...
:::    Checking for git... already installed!
:::    Checking for tar... already installed!
:::    Checking for curl... already installed!
:::    Checking for grep... already installed!
:::    Checking for dnsutils... already installed!
:::    Checking for grepcidr... not installed!
:::    Checking for whiptail... already installed!
:::    Checking for net-tools... already installed!
:::    Checking for bsdmainutils... already installed!
:::    Checking for bash-completion... already installed!
:::    Checking for iptables-persistent... already installed!
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  grepcidr
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 13.7 kB of archives.
After this operation, 29.7 kB of additional disk space will be used.
Get:1 https://deb.debian.org/debian bullseye/main armhf grepcidr armhf 2.0-2 [13.7 kB]
Fetched 13.7 kB in 0s (46.7 kB/s)
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package grepcidr.
(Reading database ... 23372 files and directories currently installed.)
Preparing to unpack .../grepcidr_2.0-2_armhf.deb ...
Unpacking grepcidr (2.0-2) ...
Setting up grepcidr (2.0-2) ...
:::    Package grepcidr successfully installed!
::: IPv6 test connections to google.com have failed. Disabling IPv6 support. (The curl test failed with code: 7)
::: Cancel selected, exiting....
```

And I end up getting the same screen: https://www.seeedstudio.com/blog/wp-content/uploads/2020/07/15936775521.png

**Reply by MichaIng** [reply]
Can you run this:
```sh
shopt -s globstar
for i in /lib/modules/**/wireguard.ko{,.xz}; do echo "$i"; done
shopt +s globstar
```

Ah, although, #1551 was just merged. But it may require some minutes before GitHub caches have been updated, when I check the script here, the above method is still there.

**Reply by simlocker** [reply]
```
root@orangedietpi:~# shopt -s globstar
in /libroot@orangedietpi:~# for i in /lib/modules/**/wireguard.ko{,.xz}; do echo "$i"; done
/lib/modules/**/wireguard.ko
/lib/modules/5.15.43-sunxi/kernel/drivers/net/wireguard/wireguard.ko.xz
root@orangedietpi:~# shopt +s globstar
-bash: shopt: +s: invalid shell option name
globstar        on
```

**Reply by coolapso** [reply]
i can see it already with the new fix when curling https://test.pivpn.io 

```
┌─[f0rs3ti@deathstar]─[~]
└──╼ $curl -SsL https://test.pivpn.io | grep wireguard.ko
	if dpkg-query -S '/lib/modules/*/wireguard.ko*' &> /dev/null || modinfo wireguard 2> /dev/null | grep -q '^filename:[[:blank:]]*(builtin)$'; then
```

maybe wait 5 more mins or so and give it another try?

**Reply by MichaIng** [reply]
So the recursive globbing for the xz-compressed variant works as expected. However, no point in debugging this now that the new method is up 🙂. If you want to test it quickly in between, the condition itself:
```sh
if dpkg-query -S '/lib/modules/*/wireguard.ko*' &> /dev/null || modinfo wireguard 2> /dev/null | grep -q '^filename:[[:blank:]]*(builtin)$'; then
  echo 'works'
else
  echo 'does not work :('
fi
```

**Reply by coolapso** [reply]
One liner version .. might be easier for copy/pasting
```
if dpkg-query -S '/lib/modules/*/wireguard.ko*' &> /dev/null || modinfo wireguard 2> /dev/null | grep -q '^filename:[[:blank:]]*(builtin)$'; then echo 'works'; else echo 'does not work :('; fi
```

**Reply by simlocker** [reply]
Do I need to do anything with that command?
Or I simply try again with "curl -L https://test.pivpn.io | TESTING= bash"?

**Reply by coolapso** [reply]
Its just a way to test the condition @MichaIng just added to the script. and check if its doing what is expected, if copy pasting what he just wrote doesn't work, just try the one i added after (its the same thing but in "one line" format).

**Reply by simlocker** [reply]
```
root@orangedietpi:~# if dpkg-query -S '/lib/modules/*/wireguard.ko*' &> /dev/null || modinfo wireguard 2> /dev/null | grep -q '^filename:[[:blank:]]*(builtin)$'; then echo 'works'; else echo 'does not work :('; fi
works
```

**Reply by coolapso** [reply]
you can go ahead and try the installation again i'd say: `curl -L https://test.pivpn.io | TESTING= bash`

**Reply by simlocker** [reply]
It is working now.

Thank you very much both.

**Reply by simlocker** [reply]
@MichaIng should I uninstall it again now and install it from the Dietpi software list? Or it won't make any difference?

**Reply by coolapso** [reply]
Awesome :D Thanks a lot @MichaIng  :D

**Reply by MichaIng** [reply]
Great! @gaston-arg you can leave it as it is, the `dietpi-software` integration does nothing more than skipping the `unattended-upgrades` installation, to not conflict with the APT auto-upgrades which can be enabled via `dietpi.txt` setting already. Otherwise it's the original installer. Ah, and until the changes got released/merged into `master`, the `dietpi-software` install still fails to allow installing WireGuard on your system 😉.

**Reply by coolapso** [reply]
Just got merged to master :) once again thanks a lot @MichaIng
